src/metabase/driver/generic_sql/query_processor.clj

```diff
@@ -40,13 +42,40 @@
     [form alias]
     form))
 
+;; TODO - Put this somewhere more generic like QP interface
+(defn- expression-with-name
+  "Return the `Expression` referenced by a given EXPRESSION-NAME."
+  [expression-name]
+  (or (some (fn [expression]
+              (when (= (name (:expression-name expression))
+                       (name expression-name))
+                expression))
+            (:expressions (:query *query*)))
+      (throw (Exception. (format "No expression named '%s'." (name expression-name))))))
 
 (defprotocol ^:private IGenericSQLFormattable
   (formatted [this]
     "Return an appropriate korma form for an object."))
 
 (extend-protocol IGenericSQLFormattable
-  nil (formatted [_] nil)
+  nil    (formatted [_] nil)
+  Number (formatted [this] this)
+  String (formatted [this] this)
+
+  Expression
+  (formatted [{:keys [operator args]}]
+    (apply (case    operator
+             :+     kx/+
+             :-     kx/-
+             :*     kx/*
+             :/     kx//
+             :lower (partial k/sqlfn* :LOWER))
+           (map formatted args)))
+
+  ExpressionRef
+  (formatted [{:keys [expression-name]}]
+    ;; Unfortunately you can't just refer to the expression by name in other clauses like filter, but have to use the original formuala.
+    (formatted (expression-with-name expression-name)))
```

Comment:
that's a possibility for SQL but non-SQL drivers will probably need to implement the functionality in a different way. Mongo, for example, will need to have a projection clause to create (and name) all the expression fields, and then refer to those elsewhere by their given name. In other words if we did it that way we'd have to write additional code to revert it back to the way it is now to use it in Mongo.